### ISSUE TITLE
Conv2DTranspose ambigous output shape

### ISSUE THREAD

**Issue by davidtvs** [keras-team/keras#9883]
Currently, ``Conv2DTranspose`` infers the shape of the output using [``deconv_length``](https://github.com/keras-team/keras/blob/af804d0a5db9a8f20fbb083b48655b2687ce89d9/keras/utils/conv_utils.py#L149) but because the output shape of a transposed convolution is ambigous it can infer an undesired shape.

Example:

```p
conv = Conv2D(16, 3, strides=2, padding='same')
transpose_conv = Conv2DTranspose(1, 3, strides=2, padding='same')

input_a = Input(shape=(23, 23, 1))
x_conv_a = conv(input_a)
x_transpose_a = transpose_conv(x_conv_a)

print("(a) Input shape: {}".format(int_shape(input_a)))
print("(a) Shape after convolution: {}".format(int_shape(x_conv_a)))
print("(a) Shape after transposed convolution: {}".format(int_shape(x_transpose_a)))
print()

input_b = Input(shape=(24, 24, 1))
x_conv_b = conv(input_b)
x_transpose_b = transpose_conv(x_conv_b)

print("(b) Input shape: {}".format(int_shape(input_b)))
print("(b) Shape after convolution: {}".format(int_shape(x_conv_b)))
print("(b) Shape after transposed convolution: {}".format(int_shape(x_transpose_b)))
```
The output:
```
(a) Input shape: (None, 23, 23, 1)
(a) Shape after convolution: (None, 12, 12, 16)
(a) Shape after transposed convolution: (None, 24, 24, 1)

(b) Input shape: (None, 24, 24, 1)
(b) Shape after convolution: (None, 12, 12, 16)
(b) Shape after transposed convolution: (None, 24, 24, 1)
```

From an input shape (None, 12, 12, 16) a transposed convolution can output either (None, 24, 24, 1) or (None, 23, 23, 1). ``Conv2DTranspose`` always outputs (None, 24, 24, 1).

Shouldn't the user have to supply the output_shape (like in [Tensorflow](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_transpose)) or an output padding (like in [PyTorch](http://pytorch.org/docs/master/nn.html?highlight=convtranspose2d#torch.nn.ConvTranspose2d)) to resolve the ambiguity?

**Comment by yangzhe151** [comment]
I also met the same question as you described, expecting someone can answer this question.

**Comment by emiljoha** [comment]
This behavior also causes serious problems when defining convolutional autoencoders where a predictable dimensionality of the output is paramount.

The backend function [conv2d_transpose](https://github.com/keras-team/keras/blob/bcf0031b54d555179be81c088cc3df0a723d7907/keras/backend/tensorflow_backend.py#L3491) that is [used](https://github.com/keras-team/keras/blob/bcf0031b54d555179be81c088cc3df0a723d7907/keras/layers/convolutional.py#L802) in the Conv2DTranspose class actually requires the desired output dimensions.

It might work to add an optional argument to specify the output dimensionality, and if used pass it to the backend. Am I missing something?

**Comment by emiljoha** [comment]
@davidtvs  Nice! Your implementation indeed seems to fix this issue! 

Why no pull request to solve the issue here in Keras?

**Comment by emiljoha** [comment]
@davidtvs  After further testing, **not** specifying `output_shape` and letting it default to `None` causes a `TypeError` exception when the `output_shape` si cast to a touple. 

`self._output_shape = tuple(output_shape)`

This is easily fixed by checking for that case and only casting to tuple if not None.

**Comment by xiaomaxiao** [comment]
@davidtvs @emiljoha  Does it support dynamic length？ eg  [None,None,3]

**Comment by emiljoha** [comment]
@xiaomaxiao Passing `[None, None, 3]` as output_shape will give you a TypeError exception. Is there a way around it? I don´t know.

**Comment by xiaomaxiao** [comment]
@davidtvs  can you fix it?

**Comment by davidtvs** [comment]
@xiaomaxiao I don't think setting `output_shape` to (None, None, 3) is going to work. Because that shape is passed directly to [`keras.backend.conv2d_transpose`](https://keras.io/backend/#conv2d_transpose) which in turn calls [`tf.nn.conv2d_transpose`](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_transpose) if you are using Tensorflow as the backend, and it doesn't accept `None` in any dimension of output shape.

The [output padding](https://github.com/davidtvs/keras/blob/17f1424f10dfd6da554c9e3715916b3a2e075876/keras/layers/convolutional.py#L621) implementation from #10246 might solve your problem though, as long as you don't need the padding to change during training. From your description in #10185, you need output_padding=0.

**Comment by xiaomaxiao** [comment]
@davidtvs   

I want set output_shape = tf.shape not (None,None,3) like this.

`x = Input([None,None,3])

output_shape = tf.shape(x)

x = Conv2D(32,kernel_size = (3,3), padding='same',strides=(2,2))(x)

x = Conv2DTranspose(ch,kernel_size=(3,3),strides=(2,2),padding='same',output_shape)`

**Comment by davidtvs** [comment]
PR #10246 has been merged. The shape of the output can now be controlled through a new optional argument (`output_padding`).
